### PR TITLE
Add prepacking for linear weights

### PR DESCRIPTION
**Pull request by Ryo-not-rio**
- add pack_linear_weight() in the python frontend
- add matmul support for mkldnn weights in the backend for prepacked weights

Context:
On aarch64 matmuls using ComputeLibrary, linear weights are reordered at runtime, adding a significant overhead to linear layers during inference. This patch allows us to do the reordering ahead of time and pack the weights meaning no reorders are needed during runtime.

Usage:
```
model = Model()
model = torch.nn.utils.pack_linear.pack_linear_weights(model)
print(model(inputs))
```

Speedups on 4 threads:
~6% for Fp32 bert base & large
~7% for autocast bert & large
~30% for fast math bert & large
~8% for fp32 clip
~4% for autocast clip
~20% for fast math clip

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @malfet @milpuz01 @EikanWang @voznesenskym @penguinwu @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0 @xmfan

### CODE REVIEW

**Comment by Ryo-not-rio** [comment]
This functionality was suggested during our PR for caching weights in ideep: https://github.com/intel/ideep/pull/331.

**Comment by vadimkantorov** [comment]
Is `COL32` and related cublas layouts the similar concept for CUBLAS?
- https://github.com/pytorch/pytorch/issues/104702

Or is it not relevant anymore in the recent versions of CUBLAS?

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -139,6 +174,39 @@ Tensor& linear_out(const Tensor& input, const Tensor& weight, const std::optiona
   return output;
 }
 
+Tensor pack_linear(const Tensor& weight) {
```
Comment (jgong5) [inline aten/src/ATen/native/Linear.cpp:177]: We have a similar aten function `mkldnn_reorder_linear_weight` (`mkldnn::_reorder_linear_weight`) in `MKLDNNConversions.cpp`. Can we use it instead? Also, since it is specific to mkldnn layout conversions, suggest to move things to `MKLDNNConversions.cpp` too.

File: aten/src/ATen/native/Linear.cpp [hunk 2]
```diff
@@ -91,7 +98,35 @@ Tensor linear(const Tensor& input, const Tensor& weight, const std::optional<Ten
     return xnnpack::linear(input, weight, *bias);
   }
 #endif
-  if (input_dim == 2 && bias->defined()) {
+
+  // Weight has been prepacked
+  if (weight.is_mkldnn()) {
```
Comment (jgong5) [inline aten/src/ATen/native/Linear.cpp:103]: Can we reuse `at::kldnn_linear` or improve it?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -6999,6 +7002,7 @@
   structured_delegate: addmm.out
   variants: function, method
   dispatch:
+    MkldnnCPU: mkldnn_addmm_cpu
```
Comment (jgong5) [inline aten/src/ATen/native/native_functions.yaml:7005]: Since we only need to handle things at the "linear" level, can we avoid changing "addmm"?

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Comment by cfRod** [comment]
@malfet

File: aten/src/ATen/native/Linear.cpp (hunk 1 above)
Comment (Ryo-not-rio) [reply aten/src/ATen/native/Linear.cpp]: Hi, thank you for the reviews. I had  a look at the `mkldnn_reorder_linear_weight` function and it doesn't look like we can use it effectively since we'd need to go to `ideep::reorder_if_differ()`. I have moved the `pack_linear` function to `MKLDNNConversions.cpp`.

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (Ryo-not-rio) [reply aten/src/ATen/native/native_functions.yaml:7005]: Unfortunately we have to change addmm since we need it to handle mkldnn layout weights now

File: aten/src/ATen/native/Linear.cpp (hunk 2 above)
Comment (jgong5) [reply aten/src/ATen/native/Linear.cpp:103]: Then, does it make sense to revise `mkldnn_linear` to use the matmul primitive instead? cc @yanbing-j 
My concern is the complexity added to the addmm, which seems not necessary since working from `linear` sounds good enough to me. If "inner_product" is currently slower, you can perhaps add a dedicated path using matmul primitives for aarch64 in `mkldnn_linear`. Later on, we can further simplify the implementation of `mkldnn_linear` to always use the matmul primitive, suppose it is also faster on x86.

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (jgong5) [reply aten/src/ATen/native/native_functions.yaml:7005]: See my comment in another hunk.

File: aten/src/ATen/native/Linear.cpp (hunk 1 above)
Comment (jgong5) [reply aten/src/ATen/native/Linear.cpp]: Still, I guess the major difference is the usage of inner product primitive vs. matmul primitive. Can we add a dedicated path for aarch64 in `mkldnn_reorder_linear_weight` for now just like the suggestion I gave for `mkldnn_linear`? We can simplify things later on. cc @yanbing-j

**Review by jgong5 — COMMENTED** [review]
Suggest we just change linear, not addmm.

File: aten/src/ATen/native/Linear.cpp (hunk 2 above)
Comment (Ryo-not-rio) [reply aten/src/ATen/native/Linear.cpp:103]: Done.

File: aten/src/ATen/native/Linear.cpp (hunk 1 above)
Comment (Ryo-not-rio) [reply aten/src/ATen/native/Linear.cpp]: Done

**Comment by eellison** [comment]
How is the different than torch.utils.mkldnn.to_mkldnn?

**Review by albanD — DISMISSED** [review]
The new public API looks quite surprising to me. This would definitely need to be approved by @mikaylagawarecki before landing within nn/

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -940,7 +940,7 @@
 - func: as_strided(Tensor(a) self, SymInt[] size, SymInt[] stride, SymInt? storage_offset=None) -> Tensor(a)
   variants: function, method
   dispatch:
-    ZeroTensor, CPU, CUDA: as_strided_tensorimpl
+    ZeroTensor, CPU, CUDA, MkldnnCPU: as_strided_tensorimpl
```
Comment (jgong5) [inline aten/src/ATen/native/native_functions.yaml:943]: mkldnn tensor is opaque without the definition of "stride". It is weird to provided an implementation as `as_strided` for it. It must be something wrong from the caller.

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp [hunk 5]
```diff
@@ -256,9 +257,41 @@ static Tensor mkldnn_reorder_conv_weight(
   }
 }
 
-static Tensor mkldnn_reorder_linear_weight(
+Tensor mkldnn_reorder_linear_weight(
     const Tensor& self,
     std::optional<int64_t> batch_size_opt) {
+
+  ideep::tensor result;
+#if defined(__aarch64__) && AT_MKLDNN_ACL_ENABLED()
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:265]: can you add a note here explaining why you cannot reuse existing implementation but have to create a special path for aarch64? We need a TODO to consolidate the implementation in the future.

File: aten/src/ATen/native/Linear.cpp [hunk 6]
```diff
@@ -62,7 +64,12 @@ static inline Tensor _flatten_nd_linear(const Tensor& input, const Tensor& weigh
       flattened_dim = flattened_dim * input_sizes[i];
     }
     auto inp_reshape = input.reshape_symint({flattened_dim, input_sizes.at(input_sizes.size() -1)});
-    const auto result = at::addmm(bias, inp_reshape, weight.t());
+    Tensor result;
+    if (weight.is_mkldnn()) {
+      result = at::addmm(bias, inp_reshape, weight);
```
Comment (jgong5) [inline aten/src/ATen/native/Linear.cpp:69]: addmm supports mkldnn cpu tensor?

File: aten/src/ATen/native/mkldnn/Linear.cpp [hunk 7]
```diff
@@ -59,6 +61,16 @@ namespace at::native {
 Tensor mkldnn_linear(
     const Tensor& self,
     const Tensor& weight_t, const std::optional<Tensor>& bias_opt) {
+#if defined(__aarch64__) && AT_MKLDNN_ACL_ENABLED()
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Linear.cpp:64]: ditto: explain why we implement specific path for aarch64.

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Comment by Ryo-not-rio** [comment]
> How is the different than torch.utils.mkldnn.to_mkldnn?

@eellison This function will not only change the weights to mkldnn tensors but also reorder the weights to the format expected by compute library

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (Ryo-not-rio) [reply aten/src/ATen/native/native_functions.yaml]: Removed

File: aten/src/ATen/native/mkldnn/Linear.cpp (hunk 7 above)
Comment (Ryo-not-rio) [reply aten/src/ATen/native/mkldnn/Linear.cpp:67]: Done

File: aten/src/ATen/native/Linear.cpp (hunk 6 above)
Comment (Ryo-not-rio) [reply aten/src/ATen/native/Linear.cpp]: Removed

File: aten/src/ATen/native/Linear.cpp [hunk 8]
```diff
@@ -62,7 +64,7 @@ static inline Tensor _flatten_nd_linear(const Tensor& input, const Tensor& weigh
       flattened_dim = flattened_dim * input_sizes[i];
     }
     auto inp_reshape = input.reshape_symint({flattened_dim, input_sizes.at(input_sizes.size() -1)});
-    const auto result = at::addmm(bias, inp_reshape, weight.t());
+    Tensor result = at::addmm(bias, inp_reshape, weight.t());
```
Comment (jgong5) [inline aten/src/ATen/native/Linear.cpp:67]: This change is not needed.

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/Linear.cpp (hunk 8 above)
Comment (Ryo-not-rio) [reply aten/src/ATen/native/Linear.cpp]: Fixed

**Review by jgong5 — APPROVED** [review]

**Comment by jerryzh168** [comment]
why does this happen in eager mode instead of other flows like pt2e quant flow?

**Comment by Ryo-not-rio** [comment]
> why does this happen in eager mode instead of other flows like pt2e quant flow?

@jerryzh168 This happens any time the GEMM kernels in ComputeLibrary is called for aarch64 through oneDNN

**Comment by Ryo-not-rio** [comment]
@jgong5 I've had to move the `pack_linear` function away from `mkldnn_reorder_linear_weights` again as the new packing function isn't compatible with the `torch.compile` path. It could be work done in the future to consolidate the two but for now I think it's best that the two remain separate. The `mkldnn_linear` function can remain as it is but I've guarded it to check for packed weights.

**Comment by Ryo-not-rio** [comment]
@mikaylagawarecki please have a look at this PR, it adds a weight packing function to the `nn` module

**Comment by jgong5** [comment]
> @jgong5 I've had to move the `pack_linear` function away from `mkldnn_reorder_linear_weights` again as the new packing function isn't compatible with the `torch.compile` path. It could be work done in the future to consolidate the two but for now I think it's best that the two remain separate. The `mkldnn_linear` function can remain as it is but I've guarded it to check for packed weights.

Can you avoid force-push? It is hard to review the incremental changes.

**Review by nikhil-arm — APPROVED** [review]
LGTM

**Review by albanD — CHANGES_REQUESTED** [review]
This needs approval from an nn maintainer before being acceptable as new public API there.

**Comment by ng-05** [comment]
> This needs approval from an nn maintainer before being acceptable as new public API there.

Hello @albanD , who is the nn maintainer? Can you please ask them to review this change as this has been long pending on meta team for a review.

File: torch/nn/utils/pack_linear.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+# Copyright 2024 Arm Limited and/or its affiliates <open-source-office@arm.com>
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/nn/utils/pack_linear.py:2]: I think it's fine to put here since this seems to be packing the weight for unquantized linear

although I'm not sure why we need the observer hooks etc. here, can you explain when are these needed @Ryo-not-rio  ?

File: test/nn/test_pack_linear.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+            Foo(), {torch.nn.Linear}, dtype=torch.qint8
+        )
+        packed_model = torch.nn.utils.pack_linear.pack_linear_weights(model)
+        inputs = (
+            torch.randint(0, 128, (batch_size, channels, M, N), dtype=torch.float32)
+            if batch_size
+            else torch.randint(0, 128, (channels, M, N), dtype=torch.float32)
+        )
+
+        self.assertEqual(model(inputs), packed_model(inputs))
```
Comment (jerryzh168) [inline test/nn/test_pack_linear.py:232]: why do we expect the result to be the same as dynamic quant?

File: torch/nn/utils/pack_linear.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ryo-not-rio) [reply torch/nn/utils/pack_linear.py:2]: I think we want to keep the observer hooks and the devices after packing the weights. The quantized packing seems to do this so I put it here as well

**Review by albanD — CHANGES_REQUESTED** [review]
Hey,
From further discussion with nn maintainers, I don't think this PR makes much sense as is.
In particular, while the speed number look great, I really doubt that anyone is going to discover this new API and thus use it.
We recommend integrating the user-facing API with the existing architecture optimization techniques (in core or ao).
The backend change could make sense, but we would need to understand how it gets exposed to the user to have confidence that this is the right way to do it.

**Comment by Ryo-not-rio** [comment]
@albanD I have adjusted the functionality now so that the weights are automatically packed for aarch64 whenever the user calls the linear layer with `torch.no_grad()`. Do let me know of what the team thinks of this approach, thanks!

File: torch/nn/modules/linear.py [hunk 11]
```diff
@@ -121,7 +122,30 @@ def reset_parameters(self) -> None:
             bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0
             init.uniform_(self.bias, -bound, bound)
 
+    def pack_weights(self, input: Tensor) -> Tensor:
```
Comment (mikaylagawarecki) [inline torch/nn/modules/linear.py:125]: Hi @Ryo-not-rio 

The newly proposed API does not address the original concern of adding MKLDNN specific APIs to the `torch.nn` python frontend.

We strongly believe that this is an architecture optimization concept andshould live with other architecture optimization systems + reuse existing mkldnn-specific architecture. For example, why can we not reuse a similar existing mkldnn linear reordering logic here

https://github.com/pytorch/pytorch/blob/a653a13ba981349c308518a6fde2b56bc76f1649/aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp#L260

Comment (Ryo-not-rio) [reply torch/nn/modules/linear.py]: @mikaylagawarecki the `pack_weights` and `unpack_weights` functions here weren't intended as an API but just rather internal functions to be used in the forward() function of the linear layer. I have changed it to avoid further confusion but with this new implementation, there are no additional APIs added. The functionality is now all contained in the forward pass of the linear layer

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp [hunk 12]
```diff
@@ -280,6 +281,48 @@ static Tensor mkldnn_reorder_linear_weight(
   return new_with_itensor_mkldnn(std::move(result), optTypeMetaToScalarType(self.options().dtype_opt()), self.options().device_opt());
 }
 
+Tensor pack_linear(const Tensor& weight, const Tensor& input) {
+#if !defined(__aarch64__) || !AT_MKLDNN_ACL_ENABLED()
```
Comment (malfet) [inline aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:285]: Can you please explain, why it needs to be limited to aarch64 and ACL instead of enabled whenever mkldnn is used?

File: test/nn/test_pack_linear.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
+#!/usr/bin/env python3
+# Copyright 2024 Arm Limited and/or its affiliates <open-source-office@arm.com>
```
Comment (malfet) [inline test/nn/test_pack_linear.py:2]: Nit: we usually don't do top level copyrights, as this one is already covered by 
https://github.com/pytorch/pytorch/blob/001ad5bef521aa92ae8209279fa549e3f46c0ecb/LICENSE#L39-L40

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp [hunk 14]
```diff
@@ -280,6 +281,48 @@ static Tensor mkldnn_reorder_linear_weight(
...
+  return weight;
+#endif
+  const auto weight_dim = weight.dim();
```
Comment (malfet) [inline aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:288]: Don't you need to check there that tensor is on CPU?

File: torch/nn/modules/linear.py [hunk 15]
```diff
@@ -122,6 +123,27 @@ def reset_parameters(self) -> None:
...
+            if torch._C._has_mkldnn:
+                packed_weight = torch._C._nn.pack_linear(self.weight, input)
+                if packed_weight is not None:
+                    self.weight = torch.nn.Parameter(packed_weight)
+            self.packed = True
+
+        if torch.is_grad_enabled() and self.packed:
+            # Unpack weights
+            if torch._C._has_mkldnn:
+                if self.weight.is_mkldnn:
```
Comment (malfet) [inline torch/nn/modules/linear.py:137]: And what end user is supposed to do, if they load serialized model with pre-packed weights, but want to unpack it on the system that does not have mkldnn enabled?

**Review by malfet — CHANGES_REQUESTED** [review]
Agree with two other reviewers, that this indeed demonstrates significant performance gains but end user experience/current implementation is problematic:
 - It will crash if one tries to run any inference on Tegra/Grace system when tensor is on GPU
 - Even though implementation does not seem to rely on generic OneDNN calls, it singles out aarch64 and skips similar path for non-ARM systems
 - It's unclear why one needs new `native_pack` API instead of just casting tensor to(mkldnn) with some specific layout

I need to check, if there is an explicit guarantee documented somewhere, it it feels intuitive that forward method in no_grad mode should not change model state, but it does after this change, without any way to opt-out of the change.

And if possible, before attaching priority label to the PR please make sure that at least linter passes locally

**Comment by Ryo-not-rio** [comment]
Closing as we are abandoning this feature
